core/src/avm2/globals/flash/net/sharedobject.rs

```diff
@@ -39,28 +45,221 @@ fn class_init<'gc>(
+    // Remove leading/trailing slashes.
+    movie_path = movie_path.strip_prefix('/').unwrap_or(movie_path);
+    movie_path = movie_path.strip_suffix('/').unwrap_or(movie_path);
+
+    let movie_host = if movie_url.scheme() == "file" {
+        // Remove drive letter on Windows (TODO: move this logic into DiskStorageBackend?)
+        if let [_, b':', b'/', ..] = movie_path.as_bytes() {
+            movie_path = &movie_path[3..];
+        }
+        "localhost"
+    } else {
+        movie_url.host_str().unwrap_or_default()
+    };
+
+    let local_path = if let Some(Value::String(local_path)) = args.get(1) {
+        // Empty local path always fails.
+        if local_path.is_empty() {
+            return Ok(Value::Null);
+        }
+
+        // Remove leading/trailing slashes.
+        let mut local_path = local_path.to_utf8_lossy();
+        if local_path.ends_with('/') {
+            match &mut local_path {
+                Cow::Owned(p) => {
+                    p.pop();
+                }
+                Cow::Borrowed(p) => *p = &p[..p.len() - 1],
+            }
+        }
+        if local_path.starts_with('/') {
+            match &mut local_path {
+                Cow::Owned(p) => {
+                    p.remove(0);
+                }
+                Cow::Borrowed(p) => *p = &p[1..],
+            }
+        }
+
+        // Verify that local_path is a prefix of the SWF path.
+        if movie_path.starts_with(local_path.as_ref())
+            && (local_path.is_empty()
+                || movie_path.len() == local_path.len()
+                || movie_path[local_path.len()..].starts_with('/'))
+        {
+            local_path
+        } else {
+            log::warn!("SharedObject.get_local: localPath parameter does not match SWF path");
+            return Ok(Value::Null);
+        }
+    } else {
+        Cow::Borrowed(movie_path)
+    };
+
+    // Final SO path: foo.com/folder/game.swf/SOName
+    // SOName may be a path containing slashes. In this case, prefix with # to mimic Flash Player behavior.
+    let prefix = if name.contains('/') { "#" } else { "" };
+    let full_name = format!("{}/{}/{}{}", movie_host, local_path, prefix, name);
+
+    // Avoid any paths with `..` to prevent SWFs from crawling the file system on desktop.
+    // Flash will generally fail to save shared objects with a path component starting with `.`,
+    // so let's disallow them altogether.
+    if full_name.split('/').any(|s| s.starts_with('.')) {
+        log::error!("SharedObject.get_local: Invalid path with .. segments");
+        return Ok(Value::Null);
+    }
+
+    // Check if this is referencing an existing shared object
+    if let Some(so) = activation.context.avm2_shared_objects.get(&full_name) {
+        return Ok((*so).into());
+    }
+
+    // Data property only should exist when created with getLocal/Remote
+    let constructor = activation.avm2().classes().sharedobject;
+    let mut this = constructor.construct(activation, &[])?;
```

Comment:
Since this path logic is fairly complex and mostly the same between AVM1 and AVM2, I wonder if we can share this in some way?

Reply:
@Herschel: The tricky part is that this uses `Value` a lot, and we don't have any API for abstracting over the AVM1 vs AVM2 `Value`.